### ISSUE TITLE
Dataset crawler crashing on dataset that has been renamed

### ISSUE THREAD

**Issue by markbrough** [codeforIATI/iati-datastore#228]
### Describe the bug

Dataset crawler critical error on dataset `dmru-998`.

```
sqlalchemy.orm.exc.FlushError: New instance <Dataset at 0x7fce4c579be0> with identity key (<class 'iatilib.model.Dataset'>, ('dmru-998',), None) conflicts with persistent instance <Dataset at 0x7fce4d0a6c18>
```

Interestingly, when navigating to the following page:
https://iatiregistry.org/dataset/dmru-998

the Registry redirects to this page:
https://iatiregistry.org/dataset/cku-998

... which is deleted.

In `__iatikitcache__/registry/metadata/`, the folder `cku` exists, which includes the file `dmru-998.json`

### Additional context

Stack trace
```
Jan 27 03:01:27 codeforiati iati[1059]: Traceback (most recent call last):
Jan 27 03:01:27 codeforiati iati[1059]:   File "/home/admin/iati-datastore/pyenv/lib/python3.6/site-packages/rq/worker.py", line 975, in perform_job
Jan 27 03:01:27 codeforiati iati[1059]:     rv = job.perform()
Jan 27 03:01:27 codeforiati iati[1059]:   File "/home/admin/iati-datastore/pyenv/lib/python3.6/site-packages/flask_rq2/job.py", line 41, in perform
Jan 27 03:01:27 codeforiati iati[1059]:     return super(FlaskJob, self).perform()
Jan 27 03:01:27 codeforiati iati[1059]:   File "/home/admin/iati-datastore/pyenv/lib/python3.6/site-packages/rq/job.py", line 696, in perform
Jan 27 03:01:27 codeforiati iati[1059]:     self._result = self._execute()
Jan 27 03:01:27 codeforiati iati[1059]:   File "/home/admin/iati-datastore/pyenv/lib/python3.6/site-packages/rq/job.py", line 719, in _execute
Jan 27 03:01:27 codeforiati iati[1059]:     return self.func(*self.args, **self.kwargs)
Jan 27 03:01:27 codeforiati iati[1059]:   File "/home/admin/iati-datastore/iati_datastore/iatilib/crawler.py", line 409, in download_and_update
Jan 27 03:01:27 codeforiati iati[1059]:     update_registry()
Jan 27 03:01:27 codeforiati iati[1059]:   File "/home/admin/iati-datastore/iati_datastore/iatilib/crawler.py", line 435, in update_registry
Jan 27 03:01:27 codeforiati iati[1059]:     datasets = fetch_dataset_list()
Jan 27 03:01:27 codeforiati iati[1059]:   File "/home/admin/iati-datastore/iati_datastore/iatilib/crawler.py", line 48, in fetch_dataset_list
Jan 27 03:01:27 codeforiati iati[1059]:     db.session.commit()
Jan 27 03:01:27 codeforiati iati[1059]:   File "/home/admin/iati-datastore/pyenv/lib/python3.6/site-packages/sqlalchemy/orm/scoping.py", line 163, in do
Jan 27 03:01:27 codeforiati iati[1059]:     return getattr(self.registry(), name)(*args, **kwargs)
Jan 27 03:01:27 codeforiati iati[1059]:   File "/home/admin/iati-datastore/pyenv/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 1042, in commit
Jan 27 03:01:27 codeforiati iati[1059]:     self.transaction.commit()
Jan 27 03:01:27 codeforiati iati[1059]:   File "/home/admin/iati-datastore/pyenv/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 504, in commit
Jan 27 03:01:27 codeforiati iati[1059]:     self._prepare_impl()
Jan 27 03:01:27 codeforiati iati[1059]:   File "/home/admin/iati-datastore/pyenv/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 483, in _prepare_impl
Jan 27 03:01:27 codeforiati iati[1059]:     self.session.flush()
Jan 27 03:01:27 codeforiati iati[1059]:   File "/home/admin/iati-datastore/pyenv/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 2536, in flush
Jan 27 03:01:27 codeforiati iati[1059]:     self._flush(objects)
Jan 27 03:01:27 codeforiati iati[1059]:   File "/home/admin/iati-datastore/pyenv/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 2678, in _flush
Jan 27 03:01:27 codeforiati iati[1059]:     transaction.rollback(_capture_exception=True)
Jan 27 03:01:27 codeforiati iati[1059]:   File "/home/admin/iati-datastore/pyenv/lib/python3.6/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
Jan 27 03:01:27 codeforiati iati[1059]:     with_traceback=exc_tb,
Jan 27 03:01:27 codeforiati iati[1059]:   File "/home/admin/iati-datastore/pyenv/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 182, in raise_
Jan 27 03:01:27 codeforiati iati[1059]:     raise exception
Jan 27 03:01:27 codeforiati iati[1059]:   File "/home/admin/iati-datastore/pyenv/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 2638, in _flush
Jan 27 03:01:27 codeforiati iati[1059]:     flush_context.execute()
Jan 27 03:01:27 codeforiati iati[1059]:   File "/home/admin/iati-datastore/pyenv/lib/python3.6/site-packages/sqlalchemy/orm/unitofwork.py", line 422, in execute
Jan 27 03:01:27 codeforiati iati[1059]:     rec.execute(self)
Jan 27 03:01:27 codeforiati iati[1059]:   File "/home/admin/iati-datastore/pyenv/lib/python3.6/site-packages/sqlalchemy/orm/unitofwork.py", line 589, in execute
Jan 27 03:01:27 codeforiati iati[1059]:     uow,
Jan 27 03:01:27 codeforiati iati[1059]:   File "/home/admin/iati-datastore/pyenv/lib/python3.6/site-packages/sqlalchemy/orm/persistence.py", line 213, in save_obj
Jan 27 03:01:27 codeforiati iati[1059]:     ) in _organize_states_for_save(base_mapper, states, uowtransaction):
Jan 27 03:01:27 codeforiati iati[1059]:   File "/home/admin/iati-datastore/pyenv/lib/python3.6/site-packages/sqlalchemy/orm/persistence.py", line 408, in _organize_states_for_save
Jan 27 03:01:27 codeforiati iati[1059]:     % (state_str(state), instance_key, state_str(existing))
Jan 27 03:01:27 codeforiati iati[1059]: sqlalchemy.orm.exc.FlushError: New instance <Dataset at 0x7fce4c579be0> with identity key (<class 'iatilib.model.Dataset'>, ('dmru-998',), None) conflicts with persistent instance <Dataset at 0x7fce4d0a6c18>
```

**Comment by markbrough** [comment]
This is the instance in the database, under `Dataset`

column | value
--- | ---
name          | dmru-998
first_seen    | 2020-12-19 01:33:03.886032
last_seen     | 2021-01-21 00:45:28
last_modified | 2020-08-27 12:11:59.289335
license       | cc-zero
is_open       | t
publisher     | dmru

**Comment by markbrough** [comment]
This is the contents of `__iatikitcache__/registry/metadata/dmru-998.json`

```json
{
    "license_title": "Creative Commons CCZero",
    "maintainer": null,
    "relationships_as_object": [],
    "private": false,
    "maintainer_email": null,
    "num_tags": 0,
    "id": "95ba744b-720a-42d9-84d4-c560132cebd3",
    "metadata_created": "2019-04-29T08:44:35.400949",
    "metadata_modified": "2020-08-27T12:11:59.289335",
    "author": null,
    "author_email": "ora@dmr.org",
    "state": "active",
    "version": null,
    "creator_user_id": "73c6f80d-e935-4ef8-9421-d35750be6d37",
    "type": "dataset",
    "resources": [
        {
            "mimetype": "",
            "cache_url": null,
            "hash": "c7437a2d66f356837d2259613b053032487d55bd",
            "description": "",
            "metadata_modified": "2020-08-27T12:11:45.293232",
            "cache_last_updated": null,
            "url": "https://aidstream.s3.us-west-2.amazonaws.com/xml/dmru-998.xml",
            "name": null,
            "format": "IATI-XML",
            "package_id": "95ba744b-720a-42d9-84d4-c560132cebd3",
            "created": "2020-08-27T12:11:45.293232",
            "state": "active",
            "mimetype_inner": null,
            "last_modified": null,
            "position": 0,
            "url_type": null,
            "id": "2ec102a9-f7f0-48c7-a458-4fb5b0309984",
            "resource_type": null,
            "size": 5905
        }
    ],
    "num_resources": 1,
    "tags": [],
    "groups": [],
    "license_id": "cc-zero",
    "relationships_as_subject": [],
    "organization": {
        "description": "",
        "created": "2018-10-22T10:26:26.501134",
        "title": "Center for Church-Based Development",
        "name": "cku",
        "is_organization": true,
        "state": "active",
        "image_url": "https://www.dmcdd.org/typo3conf/ext/provider/Resources/Public/img/dmru14/logo2.png",
        "type": "organization",
        "id": "a34a1076-b926-4da8-a00d-441a96d6c2f8",
        "approval_status": "approved"
    },
    "name": "dmru-998",
    "isopen": true,
    "url": null,
    "notes": null,
    "owner_org": "a34a1076-b926-4da8-a00d-441a96d6c2f8",
    "extras": [
        {
            "key": "activity_count",
            "value": "2"
        },
        {
            "key": "data_updated",
            "value": "2020-08-27 12:11:54"
        },
        {
            "key": "filetype",
            "value": "activity"
        },
        {
            "key": "iati_version",
            "value": "2.03"
        },
        {
            "key": "language",
            "value": "en"
        },
        {
            "key": "publisher_source_type",
            "value": "primary_source"
        },
        {
            "key": "publisher_organization_type",
            "value": "21"
        },
        {
            "key": "publisher_country",
            "value": "DK"
        },
        {
            "key": "publisher_iati_id",
            "value": "DK-CVR-12006004"
        }
    ],
    "license_url": "http://www.opendefinition.org/licenses/cc-zero",
    "title": "Danish Mission Council Development Organisation Activity File-998"
}
```

**Comment by andylolz** [comment]
I think I introduced this problem as part of crawler changes in #131. Previously, datasets were keyed on their ID. As part of that PR, I changed this to key on (publisher ID, dataset ID), just because that was convenient when working with the IATI data dump directory structure. That wasn’t a great idea! So this reverts that change, and instead we construct a big dict lookup to deal with the IATI data dump directory structure.
